lib/chewy/search/loader.rb

```diff
@@ -0,0 +1,80 @@
+module Chewy
+  module Search
+    # This class is used for two different purposes: load ORM/ODM
+    # source objects.
+    #
+    # @see Chewy::Type::Import
+    # @see Chewy::Search::Request#load
+    # @see Chewy::Search::Response#records
+    # @see Chewy::Search::Scrolling#scroll_records
+    class Loader
+      # @param indexes [Array<Chewy::Index>] list of indexes to lookup types
+      # @param only [Array<String, Symbol>] list of selected type names to load
+      # @param except [Array<String, Symbol>] list of type names which will not be loaded
+      # @param options [Hash] adapter-specific load options
+      # @see Chewy::Type::Adapter::Base#load
+      def initialize(indexes: [], only: [], except: [], **options)
+        @indexes = indexes
+        @only = Array.wrap(only).map(&:to_s)
+        @except = Array.wrap(except).map(&:to_s)
+        @options = options
+      end
+
+      # Returns a {Chewy::Type} object for index name and type name passed. Caches
+      # the result for each pair to make lookup faster.
+      #
+      # @param index [String] index name
+      # @param type [String] type name
+      # @return [Chewy::Type]
+      # @raise [NoMethodError] when index was not found
+      # @raise [Chewy::UndefinedType] when index was found, but type was not
+      def derive_type(index, type)
```

Comment:
I think `NoMethodError` is not that good explanation. Maybe it is worth to add there some meaningful one instead?